crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+ * length input, producing 32 bytes of output.
+ */
+static __owur
+int hash_h(uint8_t out[ML_KEM_PKHASH_BYTES], const uint8_t *in, size_t len,
+           mctx *ctx)
+{
+    return single_keccak(out, ML_KEM_PKHASH_BYTES, in, len,
+                         ctx->sha3_256_cache);
+}
+
+/*
+ * FIPS 203, Section 4.1, equation (4.5): G.  SHA3-512 hash of a variable
+ * length input, producing 64 bytes of output, in particular the seeds
+ * (d,z) for key generation.
+ */
+static __owur
+int hash_g(uint8_t out[ML_KEM_SEED_BYTES], const uint8_t *in, size_t len,
+           mctx *ctx)
+{
+    return single_keccak(out, ML_KEM_SEED_BYTES, in, len, ctx->sha3_512_cache);
+}
+
+/*
+ * FIPS 203, Section 4.1, equation (4.4): J. SHAKE256 taking a variable length
+ * input to compute a 32-byte implicit rejection shared secret, of the same
+ * length as the expected shared secret.  (Computed even on success to avoid
+ * side-channel leaks).
+ */
+static
+int kdf(uint8_t out[ML_KEM_SHARED_SECRET_BYTES],
+         const uint8_t z[ML_KEM_RANDOM_BYTES],
+         const uint8_t *ctext, size_t len, mctx *ctx)
+{
+    EVP_MD_CTX *mdctx;
+    int ret = 0;
+
+    /*
+     * This function's return value will be ignored, but we're not allowed to
+     * ignore the return value of EVP_DigestFinalXOF()...
+     */
+    if ((mdctx = EVP_MD_CTX_new()) != NULL
+        && EVP_DigestInit_ex(mdctx, ctx->shake256_cache, NULL)
+        && EVP_DigestUpdate(mdctx, z, ML_KEM_RANDOM_BYTES)
+        && EVP_DigestUpdate(mdctx, ctext, len)
+        && EVP_DigestFinalXOF(mdctx, out, ML_KEM_SHARED_SECRET_BYTES))
+        ret = 1;
+
+    EVP_MD_CTX_free(mdctx);
+    return ret;
+}
+
+/*
+ * FIPS 203, Section 4.2.2, Algorithm 7: SampleNTT. Rejection samples a Keccak
+ * stream to get uniformly distributed elements. This is used for matrix
+ * expansion and only operates on public inputs.
+ *
+ * The block size below needs to be a multiple of 3, but is otherwise
+ * arbitrary, the chosen block size avoids internal buffering in SHAKE128, by
+ * matching the Keccac output block size of (1600 - 256)/8 bytes, which, being
+ * 168, just happens to be a multiple of 3!
+ */
+static __owur
+int sample_scalar(scalar *out, EVP_MD_CTX *mdctx)
+{
+    int done = 0;
+    uint8_t block[168];
+    size_t i;
+    uint16_t d1, d2;
+
+    while (done < DEGREE) {
+        if (!EVP_DigestSqueeze(mdctx, block, sizeof(block)))
+            return 0;
+        for (i = 0; i < sizeof(block) && done < DEGREE; i += 3) {
+            d1 = block[i] + ((block[i + 1] & 0x0f) << 8);
+            d2 = (block[i + 1] >> 4) + (block[i + 2] << 4);
+            if (d1 < kPrime)
+                out->c[done++] = d1;
+            if (d2 < kPrime && done < DEGREE)
+                out->c[done++] = d2;
+        }
+    }
+    return 1;
+}
+
+/* reduce_once reduces 0 <= x < 2*kPrime, mod kPrime. */
+static __owur uint16_t reduce_once(uint16_t x)
+{
+    const uint16_t subtracted = x - kPrime;
+    uint16_t mask = 0u - (subtracted >> 15);
```

Comment:
I assume it is only getting away with this because it is 16 bit.. I wonder if this is going to bite us at some point.

Reply:
The data type is explicitly `uint16_t`, which is rather fundamental in ML-KEM, this does not look fragile to me. The only assumption here is that machine arithmetic is "twos complement". I am not worried about that.